src/mount/apple.rs

```diff
@@ -0,0 +1,125 @@
+        MNT_DONTBROWSE;
+        /// VFS will ignore ownership information on filesystem objects
+        MNT_IGNORE_OWNERSHIP;
+        /// filesystem was mounted by automounter
+        MNT_AUTOMOUNTED;
+        /// filesystem is journaled
+        MNT_JOURNALED;
+        /// Don't allow user extended attributes 
+        MNT_NOUSERXATTR;
+        /// filesystem should defer writes 
+        MNT_DEFWRITE;
+        /// don't block unmount if not responding
+        MNT_NOBLOCK;
+        /// file system is exported
+        MNT_EXPORTED;
+        /// All I/O to the file system should be done asynchronously.
+        MNT_ASYNC;
+        /// Force a read-write mount even if the file system appears to be
+        /// unclean.
+        MNT_FORCE;
+        /// MAC support for objects.
+        MNT_MULTILABEL;
+        /// Do not update access times.
+        MNT_NOATIME;
+        /// Disallow program execution.
+        MNT_NOEXEC;
+        /// Do not honor setuid or setgid bits on files when executing them.
+        MNT_NOSUID;
+        /// Mount read-only.
+        MNT_RDONLY;
+        /// Causes the vfs subsystem to update its data structures pertaining to
+        /// the specified already mounted file system.
+        MNT_RELOAD;
+        /// Create a snapshot of the file system.
+        ///
+        MNT_SNAPSHOT;
+        /// All I/O to the file system should be done synchronously.
+        MNT_SYNCHRONOUS;
+        /// Union with underlying fs.
+        MNT_UNION;
+        /// Indicates that the mount command is being applied to an already
+        /// mounted file system.
+        MNT_UPDATE;
+    }
+);
+
+/// Mount a file system.
+///
+/// # Arguments
+/// - `source`  -   Specifies the file system.  e.g. `/dev/sd0`.
+/// - `target` -    Specifies the destination.  e.g. `/mnt`.
+/// - `flags` -     Optional flags controlling the mount.
+/// - `data` -      Optional file system specific data.
+///
+/// # see also
+/// [`mount`](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/mount.2.html)
+pub fn mount<
+    P1: ?Sized + NixPath,
+    P2: ?Sized + NixPath,
+    P3: ?Sized + NixPath,
+>(
+    source: &P1,
+    target: &P2,
+    flags: MntFlags,
+    data: Option<&P3>,
+) -> Result<()> {
+    fn with_opt_nix_path<P, T, F>(p: Option<&P>, f: F) -> Result<T>
+    where
+        P: ?Sized + NixPath,
+        F: FnOnce(*const libc::c_char) -> T,
+    {
+        match p {
+            Some(path) => path.with_nix_path(|p_str| f(p_str.as_ptr())),
+            None => Ok(f(std::ptr::null())),
+        }
+    }
+
+    let res = source.with_nix_path(|s|
+        target.with_nix_path(|t| {
+            with_opt_nix_path(data, |d| unsafe {
+                println!("mounting {:?} to {:?}", s, t);
+                libc::mount(
+                    s.as_ptr(),
+                    t.as_ptr(),
+                    flags.bits(),
+                    d as *mut libc::c_void,
```

Comment:
For pointer casting, [we prefer `.cast()` and `.cast_mut()` over the `as` keyword](https://github.com/nix-rust/nix/blob/master/CONVENTIONS.md#pointer-type-casting):

```suggestion
                    d.cast().cast_mut(),
```

BTW, we are casting an immutable reference (`&P3`) to a muable pinter (`*mut c_void`), is this safe?

Reply:
`libc:mount()` needs to recv `libc:c_void` type, It's hard to cast type to `libc:c_void` by the `cast` function, So I have changed it to `d.cast_mut() as *mut libc::c_void,`

Reply:
>  It's hard to cast type to libc:c_void by the cast function, So I have changed it to d.cast_mut() as *mut libc::c_void,

Yeah, `.cast().cast_mut()` should not work as the pointee type is unknown, `.cast_mut().cast()` should work.

> BTW, we are casting an immutable reference (&P3) to a muable pinter (*mut c_void), is this safe?

Will `mount(2)` write to that pointer, if so, when a UB could happen:<

Reply:
From https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/mount.2.html man page
> 
     int
     mount(const char *type, const char *dir, int flags, void *data);

     int
     unmount(const char *dir, int flags);


>     
       Data is a pointer to a structure that contains the type specific argu-ments arguments
       ments to mount.  The format for these argument structures is described in
       the manual page for each filesystem.

I think it's not be written, but nobody knows how Apple doing I think.